table/pit-cs.go

```diff
@@ -80,361 +110,145 @@ type PitOutRecord struct {
-	switch csReplacementPolicy {
-	case "lru":
-		pitCs.csReplacement = NewCsLRU(pitCs)
-	default:
-		core.LogFatal(pitCs, "Unknown CS replacement policy ", csReplacementPolicy)
-	}
-	pitCs.csMap = make(map[uint64]*CsEntry)
-
-	return pitCs
-}
-
-func (p *PitCsNode) findExactMatchEntry(name *ndn.Name) *PitCsNode {
-	if name.Size() > p.depth {
-		for _, child := range p.children {
-			if name.At(child.depth - 1).Equals(child.component) {
-				return child.findExactMatchEntry(name)
-			}
-		}
-	} else if name.Size() == p.depth {
-		return p
-	}
-	return nil
-}
-
-func (p *PitCsNode) findLongestPrefixEntry(name *ndn.Name) *PitCsNode {
-	if name.Size() > p.depth {
-		for _, child := range p.children {
-			if name.At(child.depth - 1).Equals(child.component) {
-				return child.findLongestPrefixEntry(name)
-			}
-		}
-	}
-	return p
+type CsEntry interface {
+	Index() uint64 // the hash of the entry, for fast lookup
+	StaleTime() time.Time
+	Data() *ndn.Data
 }
 
-func (p *PitCsNode) fillTreeToPrefix(name *ndn.Name) *PitCsNode {
-	curNode := p.findLongestPrefixEntry(name)
-	for depth := curNode.depth + 1; depth <= name.Size(); depth++ {
-		newNode := new(PitCsNode)
-		newNode.component = name.At(depth - 1).DeepCopy()
-		newNode.depth = depth
-		newNode.parent = curNode
-		curNode.children = append(curNode.children, newNode)
-		curNode = newNode
-	}
-	return curNode
+type baseCsEntry struct {
+	index     uint64
+	staleTime time.Time
+	data      *ndn.Data
 }
 
-func (p *PitCsNode) pruneIfEmpty() {
-	for curNode := p; curNode.parent != nil && len(curNode.children) == 0 && len(curNode.pitEntries) == 0 && curNode.csEntry == nil; curNode = curNode.parent {
-		// Remove from parent's children
-		for i, child := range curNode.parent.children {
-			if child == p {
-				if i < len(curNode.parent.children)-1 {
-					copy(curNode.parent.children[i:], curNode.parent.children[i+1:])
-				}
-				curNode.parent.children = curNode.parent.children[:len(curNode.parent.children)-1]
-				break
-			}
-		}
-	}
-}
-
-func (p *PitCs) generateNewPitToken() uint32 {
-	for {
-		token := rand.Uint32()
-		if _, ok := p.pitTokenMap[token]; !ok {
-			return token
-		}
+// InsertInRecord finds or inserts an InRecord for the face, updating the metadata and returning whether there was already an in-record in the entry.
+func (bpe *basePitEntry) InsertInRecord(interest *ndn.Interest, face uint64, incomingPitToken []byte) (*PitInRecord, bool) {
+	var record *PitInRecord
+	var ok bool
+	if record, ok = bpe.inRecords[face]; !ok {
+		record := new(PitInRecord)
+		record.Face = face
+		record.LatestNonce = interest.Nonce()
+		record.LatestTimestamp = time.Now()
+		record.LatestInterest = interest
+		record.ExpirationTime = time.Now().Add(interest.Lifetime())
+		record.PitToken = incomingPitToken
+		bpe.inRecords[face] = record
+		return record, len(bpe.inRecords) > 1
```

Comment:
We only enter this scope (lines 129-139 in revised version) if `face` does not exist as a key in `bpe.inRecords`. In this case, we then set `bpe.inRecords[face] = record`. Thus, `len(bpe.inRecords) > 1` will always be false (in fact, `len(bpe.inRecords==1`). So we can probably simplify this to just always return `false`.

Reply:
Writing unit tests shows that this does seem to be a bug.

The function description is:
```
// InsertInRecord finds or inserts an InRecord for the face, updating the metadata and returning whether there was already an in-record in the entry.
```

This will only return `false` if we are inserting into an empty map. If we insert an inRecord and then insert another inRecord for a different face, it'll incorrectly return `true`. The fix is to always return `false`, which I forgot to do despite saying it.